base_layer/wallet/src/transaction_service/protocols/transaction_send_protocol.rs

```diff
@@ -394,69 +430,123 @@ where TBackend: TransactionBackend + Clone + 'static
             )
             .await
         {
-            Ok(result) => match result.resolve_ok().await {
-                Some(send_states) if send_states.len() == 1 => {
+            Ok(result) => match result {
+                SendMessageResponse::Queued(send_states) => {
+                    if self.wait_on_dail(send_states).await {
+                        direct_send_result = true;
+                    } else {
+                        store_and_forward_send_result = self.send_transaction_store_and_forward(msg.clone()).await?;
+                    }
+                },
+                SendMessageResponse::Failed => {
+                    store_and_forward_send_result = self.send_transaction_store_and_forward(msg.clone()).await?;
+                },
+                SendMessageResponse::PendingDiscovery(rx) => {
+                    store_and_forward_send_result = self.send_transaction_store_and_forward(msg.clone()).await?;
+                    // now wait for discovery to complete
+                    match rx.await {
```

Comment:
Ok to wait for discovery here because this is a task that is spawned per transaction - right?